introspection-engine/connectors/sql-introspection-connector/src/introspection.rs

```diff
@@ -83,14 +57,31 @@ pub(crate) fn introspect(ctx: &mut Context) -> Result<(Version, String, bool), S
     let version = version_checker::check_prisma_version(ctx);
 
     // if based on a previous Prisma version add id default opinionations
-    add_prisma_1_id_defaults(&version, &mut datamodel, schema, ctx);
+    add_prisma_1_id_defaults(&version, &mut datamodel, ctx.schema, ctx);
 
     let config = if ctx.render_config {
-        render_configuration(ctx.config, schema).to_string()
+        render_configuration(ctx.config, ctx.schema).to_string()
     } else {
         String::new()
     };
 
+    m2m_relations::introspect_m2m_relations(&mut datamodel, ctx);
+
+    // Ordering of model fields.
+    for model in &mut datamodel.models {
+        model
+            .fields
+            .sort_by(|a, b| match (a.as_relation_field(), b.as_relation_field()) {
+                (Some(a), Some(b)) if a.relation_info.fields.is_empty() && !b.relation_info.fields.is_empty() => {
+                    std::cmp::Ordering::Greater // back relation fields last
+                }
+                (Some(a), Some(b)) if b.relation_info.fields.is_empty() && !a.relation_info.fields.is_empty() => {
+                    std::cmp::Ordering::Less
+                }
+                _ => std::cmp::Ordering::Equal,
+            });
+    }
```

Comment:
I'll write a comment for clarification. I'd rather skip the PartialOrd here because we are going to remove the whole data structure, probably next week.